File: 0186-reverse-words-in-a-string-ii/0186-reverse-words-in-a-string-ii.js
/**
 * @param {character[]} s
 * @return {void} Do not return anything, modify s in-place instead.
 */
var reverseWords = function(s) {
    /*
        - reverse the array
        - reverse each word from start of word to end 
    */

    s.reverse()
    let start = 0;
    s.push(' ')
    for (let i = 0; i < s.length; i++) {
        if (s[i] === ' ') {
            let end = i - 1;
            while (start < end) {
                [s[start], s[end]] = [s[end], s[start]];
                start += 1;
                end -= 1;
            }
            start = i + 1;
        }
    }
    s.pop();
};

File: 0186-reverse-words-in-a-string-ii/0186-reverse-words-in-a-string-ii.py
class Solution:
    def reverseWords(self, s: List[str]) -> None:
        """
        Do not return anything, modify s in-place instead.
        """
        s.reverse()
        start = 0
        s.append(' ')
        for i in range(0, len(s)):
            if s[i] == ' ':
                end = i - 1
                while start < end:
                    s[start], s[end] = s[end], s[start]
                    start += 1
                    end -= 1
                start = i + 1
        s.pop()    